File: bookstore-pos/frontend/src/modules/pos/utils/pricing.ts
import type { CartItem } from "@/app/store";
import type { ProductPromotionRule } from "@/modules/catalog/api/promotions";

export type PackPricingLine = {
  product_id: number;
  base_total: number;
  pack_discount: number;
  final_total: number;
  applied_rule_id?: number;
  applied_rule_name?: string;
  applied_rule_type?: "BUNDLE_PRICE" | "UNIT_PRICE_BY_QTY";
  bundle_qty?: number;
  bundle_price?: number;
  bundles_applied?: number;
  min_qty?: number;
  promo_unit_price?: number;
};

export type PackPricingResult = {
  linesByProductId: Record<number, PackPricingLine>;
  grossSubtotal: number;
  subtotalAfterPacks: number;
  packDiscountTotal: number;
};

export type PosTotalsSummary = {
  subtotal: number;
  tax: number;
  total: number;
  packDiscount: number;
  promotionDiscount: number;
  totalDiscount: number;
  subtotalAfterPacks: number;
  grossSubtotal: number;
};

const calculateBundleDiscount = (qty: number, unitPrice: number, bundleQty: number, bundlePrice: number) => {
  if (qty <= 0 || unitPrice <= 0 || bundleQty <= 0 || bundlePrice <= 0) return { discount: 0, bundles: 0 };
  const bundles = Math.floor(qty / bundleQty);
  if (bundles <= 0) return { discount: 0, bundles: 0 };
  const regularBundleTotal = unitPrice * bundleQty;
  const rawBundleDiscount = regularBundleTotal - bundlePrice;
  if (rawBundleDiscount <= 0) return { discount: 0, bundles: 0 };
  const lineTotal = unitPrice * qty;
  const discount = Math.max(0, Math.min(lineTotal, bundles * rawBundleDiscount));
  return { discount, bundles };
};

const calculateUnitPriceByQtyDiscount = (qty: number, unitPrice: number, minQty: number, promoUnitPrice: number) => {
  if (qty <= 0 || unitPrice <= 0 || minQty <= 0 || promoUnitPrice <= 0) return 0;
  if (qty < minQty) return 0;
  const rawUnitDiscount = unitPrice - promoUnitPrice;
  if (rawUnitDiscount <= 0) return 0;
  const lineTotal = unitPrice * qty;
  return Math.max(0, Math.min(lineTotal, qty * rawUnitDiscount));
};

const getRuleSpecificity = (rule?: ProductPromotionRule) => {
  if (!rule) return -1;
  if (rule.rule_type === "UNIT_PRICE_BY_QTY") return Number(rule.min_qty || 0);
  return Number(rule.bundle_qty || 0);
};

const getRulePriority = (rule?: ProductPromotionRule) => {
  if (!rule) return [-1, -1] as const;
  const typePriority = rule.rule_type === "UNIT_PRICE_BY_QTY" ? 1 : 0;
  return [getRuleSpecificity(rule), typePriority] as const;
};

const selectBestRule = (item: CartItem, rules: ProductPromotionRule[]) => {
  let bestRule: ProductPromotionRule | undefined;
  let bestDiscount = 0;
  let bestBundles = 0;

  for (const rule of rules) {
    let discount = 0;
    let bundles = 0;
    if (rule.rule_type === "BUNDLE_PRICE" && rule.bundle_qty && rule.bundle_price) {
      const bundleResult = calculateBundleDiscount(item.qty, item.price, rule.bundle_qty, rule.bundle_price);
      discount = bundleResult.discount;
      bundles = bundleResult.bundles;
    } else if (rule.rule_type === "UNIT_PRICE_BY_QTY" && rule.min_qty && rule.unit_price) {
      discount = calculateUnitPriceByQtyDiscount(item.qty, item.price, rule.min_qty, rule.unit_price);
    }

    if (discount > bestDiscount) {
      bestRule = rule;
      bestDiscount = discount;
      bestBundles = bundles;
    } else if (discount === bestDiscount && discount > 0) {
      const currentPriority = getRulePriority(rule);
      const bestPriority = getRulePriority(bestRule);
      if (
        currentPriority[0] > bestPriority[0] ||
        (currentPriority[0] === bestPriority[0] && currentPriority[1] > bestPriority[1])
      ) {
        bestRule = rule;
        bestBundles = bundles;
      }
    }
  }

  return { rule: bestRule, discount: bestDiscount, bundles: bestBundles };
};

export const calculatePackPricing = (items: CartItem[], rules: ProductPromotionRule[]): PackPricingResult => {
  const rulesByProduct: Record<number, ProductPromotionRule[]> = {};
  rules
    .filter((rule) => rule.is_active)
    .forEach((rule) => {
      if (!rulesByProduct[rule.product_id]) rulesByProduct[rule.product_id] = [];
      rulesByProduct[rule.product_id].push(rule);
    });

  const linesByProductId: Record<number, PackPricingLine> = {};
  let grossSubtotal = 0;
  let subtotalAfterPacks = 0;
  let packDiscountTotal = 0;

  items.forEach((item) => {
    const baseTotal = item.price * item.qty;
    const { rule, discount, bundles } = selectBestRule(item, rulesByProduct[item.product_id] || []);
    const safeDiscount = Math.max(0, Math.min(baseTotal, discount));
    const finalTotal = baseTotal - safeDiscount;
    grossSubtotal += baseTotal;
    subtotalAfterPacks += finalTotal;
    packDiscountTotal += safeDiscount;

    const bundleMeta =
      rule?.rule_type === "BUNDLE_PRICE"
        ? {
            bundle_qty: rule.bundle_qty,
            bundle_price: rule.bundle_price,
            bundles_applied: bundles,
          }
        : {};

    const unitMeta =
      rule?.rule_type === "UNIT_PRICE_BY_QTY"
        ? {
            min_qty: rule.min_qty,
            promo_unit_price: rule.unit_price,
          }
        : {};

    linesByProductId[item.product_id] = {
      product_id: item.product_id,
      base_total: baseTotal,
      pack_discount: safeDiscount,
      final_total: finalTotal,
      applied_rule_id: rule?.id,
      applied_rule_name: rule?.name,
      applied_rule_type: rule?.rule_type,
      ...bundleMeta,
      ...unitMeta,
    };
  });

  return {
    linesByProductId,
    grossSubtotal,
    subtotalAfterPacks,
    packDiscountTotal,
  };
};

export const calculatePosTotalsSummary = (args: {
  grossSubtotal: number;
  subtotalAfterPacks: number;
  packDiscount: number;
  promotionDiscount: number;
  taxRate: number;
  taxIncluded: boolean;
}): PosTotalsSummary => {
  const subtotalAfterPacks = Math.max(0, args.subtotalAfterPacks);
  const packDiscount = Math.max(0, args.packDiscount);
  const promotionDiscount = Math.max(0, Math.min(subtotalAfterPacks, args.promotionDiscount));
  const rate = Number(args.taxRate) || 0;

  let subtotal = subtotalAfterPacks;
  let tax = 0;
  let total = 0;
  if (args.taxIncluded) {
    if (rate > 0) {
      tax = subtotalAfterPacks - subtotalAfterPacks / (1 + rate / 100);
    }
    subtotal = subtotalAfterPacks - tax;
    total = subtotalAfterPacks - promotionDiscount;
  } else {
    subtotal = subtotalAfterPacks;
    tax = subtotal * (rate / 100);
    total = subtotal + tax - promotionDiscount;
  }
  total = Math.max(0, total);

  return {
    grossSubtotal: Math.max(0, args.grossSubtotal),
    subtotalAfterPacks,
    subtotal,
    tax,
    total,
    packDiscount,
    promotionDiscount,
    totalDiscount: packDiscount + promotionDiscount,
  };
};

File: bookstore-pos/backend/app/services/pos/pricing.py
from dataclasses import dataclass
from typing import Literal

RuleType = Literal["BUNDLE_PRICE", "UNIT_PRICE_BY_QTY"]


@dataclass(frozen=True)
class BundleRuleInput:
    id: int
    name: str
    bundle_qty: int
    bundle_price: float


@dataclass(frozen=True)
class ProductRuleInput:
    id: int
    name: str
    rule_type: RuleType
    bundle_qty: int | None = None
    bundle_price: float | None = None
    min_qty: int | None = None
    unit_price: float | None = None
    priority: int = 0


@dataclass(frozen=True)
class BundlePricingResult:
    discount: float
    bundles_applied: int
    applied_rule: BundleRuleInput | None


@dataclass(frozen=True)
class ProductRulePricingResult:
    discount: float
    bundles_applied: int
    applied_rule: ProductRuleInput | None


@dataclass(frozen=True)
class CartLineInput:
    product_id: int
    quantity: int
    original_unit_price: float


@dataclass(frozen=True)
class AppliedPromotionLine:
    product_id: int
    quantity: int
    original_unit_price: float
    final_unit_price: float
    promotion_applied: bool
    promotion_name: str | None
    line_subtotal: float
    discount: float
    applied_rule_id: int | None
    applied_rule_type: RuleType | None
    bundles_applied: int = 0


def calculate_bundle_discount(qty: int, unit_price: float, bundle_qty: int, bundle_price: float) -> tuple[float, int]:
    if qty <= 0 or unit_price <= 0 or bundle_qty <= 0 or bundle_price <= 0:
        return 0.0, 0

    bundles = qty // bundle_qty
    if bundles <= 0:
        return 0.0, 0

    regular_bundle_total = unit_price * bundle_qty
    raw_bundle_discount = regular_bundle_total - bundle_price
    if raw_bundle_discount <= 0:
        return 0.0, 0

    line_total = unit_price * qty
    discount = bundles * raw_bundle_discount
    discount = max(0.0, min(discount, line_total))
    return float(discount), bundles


def calculate_unit_price_by_qty_discount(qty: int, unit_price: float, min_qty: int, promo_unit_price: float) -> float:
    if qty <= 0 or unit_price <= 0 or min_qty <= 0 or promo_unit_price <= 0:
        return 0.0
    if qty < min_qty:
        return 0.0

    raw_unit_discount = unit_price - promo_unit_price
    if raw_unit_discount <= 0:
        return 0.0

    line_total = unit_price * qty
    discount = qty * raw_unit_discount
    discount = max(0.0, min(discount, line_total))
    return float(discount)


def select_best_bundle_rule(qty: int, unit_price: float, rules: list[BundleRuleInput]) -> BundlePricingResult:
    best_rule: BundleRuleInput | None = None
    best_discount = 0.0
    best_bundles = 0

    for rule in rules:
        discount, bundles = calculate_bundle_discount(qty, unit_price, rule.bundle_qty, rule.bundle_price)
        if discount > best_discount:
            best_discount = discount
            best_bundles = bundles
            best_rule = rule

    return BundlePricingResult(discount=float(best_discount), bundles_applied=best_bundles, applied_rule=best_rule)


def _rule_specificity(rule: ProductRuleInput | None) -> int:
    if not rule:
        return -1
    if rule.rule_type == "UNIT_PRICE_BY_QTY":
        return int(rule.min_qty or 0)
    return int(rule.bundle_qty or 0)


def _rule_priority(rule: ProductRuleInput | None) -> tuple[int, int, int]:
    if not rule:
        return -1, -1, -1
    type_priority = 1 if rule.rule_type == "UNIT_PRICE_BY_QTY" else 0
    return int(rule.priority or 0), _rule_specificity(rule), type_priority


def select_best_product_rule(qty: int, unit_price: float, rules: list[ProductRuleInput]) -> ProductRulePricingResult:
    best_rule: ProductRuleInput | None = None
    best_discount = 0.0
    best_bundles = 0

    for rule in rules:
        discount = 0.0
        bundles_applied = 0
        if rule.rule_type == "BUNDLE_PRICE" and rule.bundle_qty and rule.bundle_price:
            discount, bundles_applied = calculate_bundle_discount(qty, unit_price, rule.bundle_qty, rule.bundle_price)
        elif rule.rule_type == "UNIT_PRICE_BY_QTY" and rule.min_qty and rule.unit_price:
            discount = calculate_unit_price_by_qty_discount(qty, unit_price, rule.min_qty, rule.unit_price)

        if discount > best_discount:
            best_discount = discount
            best_bundles = bundles_applied
            best_rule = rule
        elif discount == best_discount and discount > 0 and _rule_priority(rule) > _rule_priority(best_rule):
            best_bundles = bundles_applied
            best_rule = rule

    return ProductRulePricingResult(discount=float(best_discount), bundles_applied=best_bundles, applied_rule=best_rule)


def apply_product_promotions(
    cart_items: list[CartLineInput],
    rules_by_product: dict[int, list[ProductRuleInput]],
) -> list[AppliedPromotionLine]:
    lines: list[AppliedPromotionLine] = []
    for item in cart_items:
        qty = max(0, int(item.quantity))
        base_price = max(0.0, float(item.original_unit_price))
        base_total = base_price * qty

        result = select_best_product_rule(qty, base_price, rules_by_product.get(item.product_id, []))
        discount = max(0.0, min(base_total, float(result.discount)))
        line_subtotal = base_total - discount
        final_unit_price = (line_subtotal / qty) if qty > 0 else 0.0

        applied_rule = result.applied_rule if discount > 0 else None
        lines.append(
            AppliedPromotionLine(
                product_id=item.product_id,
                quantity=qty,
                original_unit_price=base_price,
                final_unit_price=float(final_unit_price),
                promotion_applied=applied_rule is not None,
                promotion_name=applied_rule.name if applied_rule else None,
                line_subtotal=float(line_subtotal),
                discount=float(discount),
                applied_rule_id=applied_rule.id if applied_rule else None,
                applied_rule_type=applied_rule.rule_type if applied_rule else None,
                bundles_applied=result.bundles_applied if applied_rule and applied_rule.rule_type == "BUNDLE_PRICE" else 0,
            )
        )
    return lines
